airbyte-integrations/connectors/destination-clickhouse-v2/src/main/kotlin/io/airbyte/integrations/destination/clickhouse_v2/write/load/BinaryRowInsertBuffer.kt

```diff
@@ -37,20 +53,48 @@ private val log = KotlinLogging.logger {}
  */
 @SuppressFBWarnings(
     value = ["NP_NONNULL_PARAM_VIOLATION"],
-    justification = "suspend and fb's non-null analysis don't play well"
+    justification = "suspend and fb's non-null analysis don't play well",
 )
 class BinaryRowInsertBuffer(
     val tableName: TableName,
     private val clickhouseClient: Client,
+    private val airbyteSchema: AirbyteType
 ) {
     // Initialize the inner buffer
-    private val schema = clickhouseClient.getTableSchema(tableName.name, tableName.namespace)
+    private val schema: TableSchema =
+        clickhouseClient.getTableSchema(tableName.name, tableName.namespace)
     @VisibleForTesting internal var inner = InputOutputBuffer()
     @VisibleForTesting
     internal var writer = RowBinaryFormatWriter(inner, schema, ClickHouseFormat.RowBinary)
 
+    internal val formattedSchema =
+        if (airbyteSchema is ObjectType) {
+            airbyteSchema.asColumns().mapKeys { (k, _) -> k.toClickHouseCompatibleName() }
```

Comment:
this should be using the ColumnNameMapping (from TableCatalog -> TableNameInfo -> columnNameMapping)

otherwise you're at risk of having two columns end up with the same name (which causes an error at best, and data loss at worst)

Reply:
Wouldn't that fail upon table creation though?

Reply:
TableCatalogFactory will (iirc) append a suffix to one of the columns to prevent the collision. As long as you're using the names from TableCatalog, it should work correctly

(but presumably that means that when you write data into the binary format, you need to use that suffixed column name)